### PR TITLE
Add types

### PR DESCRIPTION
**Pull request by leonardlin**
Added types to most places

### CODE REVIEW

File: lib/agenda/db-init.ts
Comment (simllll) [inline lib/agenda/db-init.ts:16]: this changes behaviour..
```
false || 'agendaJobs' results to "agendaJobs"
false ?? 'agendaJobs' results to false
```
is this intended?

File: lib/agenda/define.ts
Comment (simllll) [inline lib/agenda/define.ts:30]: processor is not exactly () => void it has two valid signatures:
`processor: ((job: Job) => Promise<void>) | ((job: Job, done) => void),`

I would suggest you make a `type Processor = processor: ((job: Job) => Promise<void>) | ((job: Job, done) => void);` and use this here twice.

File: lib/agenda/define.ts
Comment (simllll) [inline lib/agenda/define.ts:32]: `"as Processor"`

Comment (simllll) [reply lib/agenda/define.ts:32]: you could ppbl also improve the check.. e.g. by checking `typeof options === 'function'` ..

File: lib/agenda/every.ts
Comment (simllll) [inline lib/agenda/every.ts:18]: if i remember right, the interval can also be a number

File: lib/agenda/every.ts
Comment (simllll) [inline lib/agenda/every.ts:43]: instead of "any" you could use "unknown" .. which forces you to do a type check  in case you want to access a property, which is way more safe.

File: lib/agenda/index.ts
Comment (simllll) [inline lib/agenda/index.ts:31]: looks good, but you could also "split" it up like I did in the hokify fork. E.g. right now it doesn't make sense if a use provides "db" and "mongo".. by using a construct like this https://github.com/hokify/agenda/blob/c5e21e98b92ef1ba77dab36f68a331c4b493b382/src/index.ts#L85 you could achive a higher level of accuracy

see also https://github.com/hokify/agenda/blob/master/src/types/DbOptions.ts

File: lib/agenda/index.ts
Comment (simllll) [inline lib/agenda/index.ts:119]: this was also something I was experimenting on.. as right now "from outside" it looks liek there ar eno default values, and you have to check the documetation of agenda again to find out.
another way of approach this , is to set default values in the constrcutor, check out https://github.com/hokify/agenda/blob/c5e21e98b92ef1ba77dab36f68a331c4b493b382/src/index.ts#L75 to see how I did it

File: lib/agenda/index.ts
Comment (simllll) [inline lib/agenda/index.ts:124]: again here you change the behaviour slightly by using ?? instead of || ...intended? does it make sense to set a default concurrency of 0?

File: lib/agenda/index.ts
Comment (simllll) [inline lib/agenda/index.ts:144]: what's the error you get here?

File: lib/agenda/job-processing-queue.ts
Comment (simllll) [inline lib/agenda/job-processing-queue.ts:50]: nextRunAt is actually not always set, at least if i remember correctly. (never happened to me in production though), so good to go for me, but if we want to be "100%" safe, we should add an additional check like:
`if (element.attrs.nextRunAt && element.attrs.nextRunAt.getTime() <= job.attrs.nextRunAt!.getTime()) {`

File: lib/agenda/mongo.ts
Comment (simllll) [inline lib/agenda/mongo.ts:12]: why not making collection optoinal? .. `collection?: string`

File: lib/job/compute-next-run-at.ts
Comment (simllll) [inline lib/job/compute-next-run-at.ts:23]: date is Date isn't it?

File: lib/job/compute-next-run-at.ts
Comment (simllll) [inline lib/job/compute-next-run-at.ts:41]: `const cronOptions = { currentDate: lastRun, tz: timezone };` wouuld infer the correct type all time

File: lib/job/index.ts
Comment (simllll) [inline lib/job/index.ts:99]: ?? instead of || intended?

File: lib/job/remove.ts
Comment (simllll) [inline lib/job/remove.ts:8]: no need for async or? instead you could add a `:Promise<..>` as return value

File: lib/job/save.ts
Comment (simllll) [inline lib/job/save.ts:9]: no async, but add return type

File: lib/utils/process-jobs.ts
Comment (simllll) [inline lib/utils/process-jobs.ts:230]: not quite sure about this... shouldn't it be 
    `if (job.attrs.nextRunAt && job.attrs.nextRunAt <= now) {
`
to keep the logic the same? "undefined < Date" always returns false

File: lib/utils/process-jobs.ts
Comment (simllll) [inline lib/utils/process-jobs.ts:256]: same for this check, shouldn it be:
`          if (job.attrs.lockedAt && job.attrs.lockedAt < lockDeadline) {
`

**Review by simllll — COMMENTED** [review]
nice work :-) I just added some comments, most of them are just things that I remembered from refactoring it last time, most of it are nothing to worry about it though :)

File: lib/agenda/db-init.ts
Comment (leonardlin) [reply lib/agenda/db-init.ts:16]: hmm, yeah I was thinking the same thing. I was aware that they don't behave the same.
I was thinking that back in the days people used "||" but to do more precisely "??"

So while changing I checked that all the unit-tests still run fine.

I was hoping to cleanup null vs undefined un-preciseness. 
By introducing this difference, I thought that going forward we are more aware when we want to use "undefined" and when we actually want to use "null".

But I think will go through the places and see, where mixing null/undefined will cause issues and where we use '??' to express default-options.

File: lib/job/index.ts
Comment (leonardlin) [reply lib/job/index.ts:99]: it's intended here, as we check for "undefined" not "null".
I will check type definitions that 'null' is invalid.

File: lib/job/remove.ts
Comment (leonardlin) [reply lib/job/remove.ts:8]: Typescript/eslint enforces that that "async" - "Promise" come in tandem.
You can't have one without the other.

File: lib/utils/process-jobs.ts
Comment (leonardlin) [reply lib/utils/process-jobs.ts:230]: There is an "!" at the beginning.

To maintain the old logic the if - condition is.

(job.attrs.nextRunAt == falsly) OR (<= now)

File: lib/agenda/every.ts
Comment (leonardlin) [reply lib/agenda/every.ts:18]: I just checked again. "interval" is passed down further down the chain and eventually ends up in JobAttribute.repeatInterval.
This fields needs to be a "human readable" time-interval or cron-expression.

File: lib/agenda/every.ts
Comment (leonardlin) [reply lib/agenda/every.ts:43]: changed to unknown

**Comment by leonardlin** [comment]
@simllll  thanks for the review comment. I updated the PR

File: lib/agenda/index.ts
Comment (koresar) [reply lib/agenda/index.ts:133]: I'd second that.

**Review by koresar — CHANGES_REQUESTED** [review]
I am by no means a TS expert, but I have only one question really.
I'd love this merged ASAP.

In fact, I merged latest `master` to this branch and then fixed all the hiccups. FYI @leonardlin

**Comment by leonardlin** [comment]
@koresar  thanks for updating the branch. I'm eager to merge it as well but was waiting for an approver.

About "??" vs "||"

In places where we were unprecise (legacy), I left it "||"

Newer places where we want to capture "default value" I used "??"

Example case:
```
config.defaultLockLimit = 0;
this._defaultLockLimit = config.defaultLockLimit ?? 5;
// this._defaultLockLimit == 0
```
vs
```
this._defaultLockLimit = config.defaultLockLimit || 5;
// this._defaultLockLimit == 5
```

What should it be?

**Comment by koresar** [comment]
If we put ?? it would change the default, that would be a breaking change.
If the next release is v5 then it's okay. But it's rather do fewer breaking
changes in projects like this.

**Comment by leonardlin** [comment]
good point. I will check where I changed it and reverse it.

**Comment by leonardlin** [comment]
@koresar  I updated the ?? || thing. can you review

**Review by koresar — APPROVED** [review]

File: lib/agenda/define.ts
Comment (OmgImAlexis) [reply lib/agenda/define.ts]: `done` should be typed.

**Pull request merged**
